### ISSUE TITLE
[Autoscaler] AttributeError: 'NoneType' object has no attribute 'pending_nodes'

### ISSUE THREAD

**Issue by jjyao** [ray-project/ray#34870]
### What happened + What you expected to happen

```
2023-04-27 23:14:53,024 ERROR autoscaler.py:372 -- StandardAutoscaler: Error during autoscaling.
Traceback (most recent call last):
  File "/app/v2_autoscaler/autoscaler_nightly/monitor_image.binary.runfiles/py_deps_autoscaler_nightly_ray/site-packages/ray/autoscaler/_private/autoscaler.py", line 369, in update
    self._update()
  File "/app/v2_autoscaler/autoscaler_nightly/monitor_image.binary.runfiles/py_deps_autoscaler_nightly_ray/site-packages/ray/autoscaler/_private/autoscaler.py", line 392, in _update
    self.non_terminated_nodes = NonTerminatedNodes(self.provider)
  File "/app/v2_autoscaler/autoscaler_nightly/monitor_image.binary.runfiles/py_deps_autoscaler_nightly_ray/site-packages/ray/autoscaler/_private/autoscaler.py", line 120, in __init__
    self.all_node_ids = provider.non_terminated_nodes({})
  File "/app/v2_autoscaler/autoscaler_nightly/monitor_image.binary.runfiles/product/v2_autoscaler/autoscaler_nightly/node_provider_v2.py", line 240, in non_terminated_nodes
    response = self.grpc_node_provider_client.NonTerminatedNodes(request)
  File "/app/v2_autoscaler/autoscaler_nightly/monitor_image.binary.runfiles/com_github_grpc_grpc/src/python/grpcio/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/app/v2_autoscaler/autoscaler_nightly/monitor_image.binary.runfiles/com_github_grpc_grpc/src/python/grpcio/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.UNAVAILABLE
        details = "cluster is initializing by cluster manager"
        debug_error_string = "{"created":"@1682637293.023918697","description":"Error received from peer ipv4:127.0.0.1:9478","file":"external/com_github_grpc_grpc/src/core/lib/surface/call.cc","file_line":966,"grpc_message":"cluster is initializing by cluster manager","grpc_status":14}"
>
2023-04-27 23:14:53,025 ERROR monitor.py:446 -- Monitor: Execution exception. Trying again...
Traceback (most recent call last):
  File "/app/v2_autoscaler/autoscaler_nightly/monitor_image.binary.runfiles/py_deps_autoscaler_nightly_ray/site-packages/ray/autoscaler/_private/monitor.py", line 420, in _run
    len(autoscaler_summary.pending_nodes)
AttributeError: 'NoneType' object has no attribute 'pending_nodes'
```

### Versions / Dependencies

master

### Reproduction script

You can get more logs from `anyscale logs cluster --id ses_h5fx63lgseuvg3ty9gjxu6csbp --download --download-dir /tmp`

### Issue Severity

None

**Comment by samdyzon** [comment]
We're seeing a similar issue with `kuberay==0.5.0` on AKS:

```
2023-05-04 18:04:36,544	INFO monitor.py:167 -- session_name: session_2023-05-04_18-04-26_930853_8
2023-05-04 18:04:36,546	INFO monitor.py:198 -- Starting autoscaler metrics server on port 44217
2023-05-04 18:04:36,546	INFO monitor.py:218 -- Monitor: Started
2023-05-04 18:04:36,626	INFO node_provider.py:211 -- Creating KuberayNodeProvider.
2023-05-04 18:04:36,626	INFO autoscaler.py:276 -- disable_node_updaters:True
2023-05-04 18:04:36,626	INFO autoscaler.py:284 -- disable_launch_config_check:True
2023-05-04 18:04:36,626	INFO autoscaler.py:296 -- foreground_node_launch:True
2023-05-04 18:04:36,627	INFO autoscaler.py:306 -- worker_liveness_check:False
2023-05-04 18:04:36,627	INFO autoscaler.py:314 -- worker_rpc_drain:True
2023-05-04 18:04:36,627	INFO autoscaler.py:364 -- StandardAutoscaler: {'provider': {'type': 'kuberay', 'namespace': 'ray', 'disable_node_updaters': True, 'disable_launch_config_check': True, 'foreground_node_launch': True, 'worker_liveness_check': False, 'worker_rpc_drain': True}, 'cluster_name': 'ray-cluster', 'head_node_type': 'head-group', 'available_node_types': {'head-group': {'min_workers': 0, 'max_workers': 0, 'node_config': {}, 'resources': {'CPU': 0, 'memory': 2000000000}}, ...
2023-05-04 18:04:36,756	ERROR autoscaler.py:372 -- StandardAutoscaler: Error during autoscaling.
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/_private/autoscaler.py", line 369, in update
    self._update()
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/_private/autoscaler.py", line 392, in _update
    self.non_terminated_nodes = NonTerminatedNodes(self.provider)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/_private/autoscaler.py", line 120, in __init__
    self.all_node_ids = provider.non_terminated_nodes({})
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/batching_node_provider.py", line 155, in non_terminated_nodes
    self.node_data_dict = self.get_node_data()
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/_private/kuberay/node_provider.py", line 237, in get_node_data
    resource_version = self._get_pods_resource_version()
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/_private/kuberay/node_provider.py", line 358, in _get_pods_resource_version
    pod_resp = self._patch(resource_path, payload)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/_private/kuberay/node_provider.py", line 442, in _patch
    result.raise_for_status()
  File "/home/ray/anaconda3/lib/python3.9/site-packages/requests/models.py", line 1021, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 422 Client Error: Unprocessable Entity for url: https://kubernetes.default:443/api/v1/namespaces/ray/pods/ray-cluster-head-sglqp
2023-05-04 18:04:36,757	ERROR monitor.py:503 -- Error in monitor loop
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/_private/monitor.py", line 548, in run
    self._run()
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/autoscaler/_private/monitor.py", line 420, in _run
    len(autoscaler_summary.pending_nodes)
AttributeError: 'NoneType' object has no attribute 'pending_nodes'
```

I noticed that `ray.autoscaler._private.kuberay.node_provider:_get_pods_resource_version` is attempting to patch the `ray.io/autoscaler-update-timestamp` annotation on the head node, but is throwing a 422 from the kube API. Upon inspection, the head node pod does not have that annotation at all, so I wonder if the `replace` patch is throwing 422? I attempted to add the annotation in the ray cluster spec, but kuberay doesnt seem to pass this annotation down.

**Comment by wuisawesome** [comment]
ack i'll try to get a fix into master by the end of next week.

**Comment by samdyzon** [comment]
Just as an update, the root cause of my particular issue was the `azure-workload-identity` controller which was causing a large diff when attempting to update the annotations of pods that are affected by the workload identity controller. Updating to `azure-workload-identity==1.0.0` fixed the 422 coming from the kubernetes API. My issue is, of course, completely different from the issue in the original issue description, but I thought I'd at least update for anyone who may stumble upon this thread.

**Comment by wuisawesome** [comment]
Hmmm I've done some investigating and it seems that some of the proactive code cleanup in https://github.com/ray-project/ray/pull/33975 should've already fixed the issue in master actually. 

Unfortunately that commit landed just after the 2.4.0 branch was cut, so the fix is only in master and the upcoming `ray >= 2.5.0` release. 

I'm going to close this issue since it that bug should no longer be possible/that code no longer exists.
